core/src/main/java/com/oath/oak/Chunk.java

```diff
@@ -237,68 +292,111 @@ ByteBuffer readMaxKey() {
+    }
+
     /**
      * * sets the field of specified offset to 'value' for given item in entry array
+     * NOT ATOMIC!
+     * It can be used only on fields which do not change through out the entry's life, i.e. key reference, or it can
+     * be used when the entry is not yet linked.
      */
-    private void setEntryField(int item, int offset, int value) {
-        assert item + offset >= 0;
+    private void setEntryFieldInt(int item, OFFSET offset, int value) {
+        assert item + offset.value >= 0;
         switch (offset) {
-            case OFFSET_KEY_LENGTH:
+            case KEY_LENGTH:
                 // OFFSET_KEY_LENGTH and OFFSET_KEY_BLOCK should be less then 16 bits long
                 // *2 in order to get read of the signed vs unsigned limits
                 assert value < Short.MAX_VALUE * 2;
-                // set two low bytes of the handle index int
-                entries[item + OFFSET_KEY_LENGTH] =
-                        (entries[item + OFFSET_KEY_LENGTH]) | (value & KEY_LENGTH_MASK);
+                // set two low bytes of the key block id and length index
+                entries[item + offset.value] =
+                        (entries[item + offset.value]) | (value & KEY_LENGTH_MASK);
                 return;
-            case OFFSET_KEY_BLOCK:
+            case KEY_BLOCK:
                 // OFFSET_KEY_LENGTH and OFFSET_KEY_BLOCK should be less then 16 bits long
                 // *2 in order to get read of the signed vs unsigned limits
                 assert value < Short.MAX_VALUE * 2;
                 // offset must be OFFSET_KEY_BLOCK,
                 // set 2 high bytes of the int inside OFFSET_KEY_LENGTH
                 assert value > 0; // block id can never be 0
-                entries[item + OFFSET_KEY_LENGTH] =
-                        (entries[item + OFFSET_KEY_LENGTH]) | (value << KEY_BLOCK_SHIFT);
+                entries[item + offset.value] =
+                        (entries[item + offset.value]) | (value << KEY_BLOCK_SHIFT);
                 return;
-            default:
-                entries[item + offset] = value;
+            case VALUE_LENGTH:
+                // make sure the length is at most 2^23 and at least 0
+                assert (value & VALUE_LENGTH_MASK) == value;
+                entries[item + offset.value] =
+                        (entries[item + offset.value]) | (value & VALUE_LENGTH_MASK);
                 return;
+            case VALUE_BLOCK:
+                assert value > 0; // block id can never be 0
+                assert ((value << VALUE_BLOCK_SHIFT) >>> VALUE_BLOCK_SHIFT) == value; // value is up to 2^9
+                entries[item + offset.value] =
+                        (entries[item + offset.value]) | (value << VALUE_BLOCK_SHIFT);
+                return;
+            default:
+                entries[item + offset.value] = value;
         }
     }
 
+    private void setEntryFieldLong(int item, OFFSET offset, long value) {
```

Comment:
This one is indeed NOT ATOMIC :) Please comment

Reply:
I used putLongVolatile which makes it atomic